File: ros2_publisher/plc_publisher/src/plcpublisher.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "plc_msgs/msg/camera_control_custom.hpp"
#include "plc_msgs/msg/lamp_control_custom.hpp"
#include "plc_msgs/msg/pan_and_tilt_control_custom.hpp"

using namespace std::chrono_literals;

class ControlPublisher : public rclcpp::Node
{
public:
  ControlPublisher()
  : Node("control_publisher")
  {
    // Declare parameters for Camera Control
    this->declare_parameter("camera_id", 9);
    this->declare_parameter("camera_power", true);
    this->declare_parameter("camera_light", true);
    this->declare_parameter("camera_focus", 22.0);
    this->declare_parameter("camera_zoom", 25.0);

    // Declare parameters for Lamp Control
    this->declare_parameter("lamp_id", 9);
    this->declare_parameter("lamp_intensity", 80.0);
    this->declare_parameter("lamp_power", true);

    // Declare parameters for Pan and Tilt Control
    this->declare_parameter("panandtilt_id", 9);
    this->declare_parameter("panandtilt_x", 80.0);
    this->declare_parameter("panandtilt_y", 20.0);
    this->declare_parameter("panandtilt_z", 23.0);

    // Publishers for each topic
    camera_publisher_ = this->create_publisher<plc_msgs::msg::CameraControlCustom>("CameraSetterTopic", 10);
    lamp_publisher_ = this->create_publisher<plc_msgs::msg::LampControlCustom>("LampSetterTopic", 10);
    panandtilt_publisher_ = this->create_publisher<plc_msgs::msg::PanAndTiltControlCustom>("PanAndTiltSetterTopic", 10);

    // Timers for periodic publishing (1 Hz frequency)
    timer_camera_ = this->create_wall_timer(
      1s, std::bind(&ControlPublisher::publish_camera_data, this));
    timer_lamp_ = this->create_wall_timer(
      1s, std::bind(&ControlPublisher::publish_lamp_data, this));
    timer_panandtilt_ = this->create_wall_timer(
      1s, std::bind(&ControlPublisher::publish_panandtilt_data, this));
  }

private:
  void publish_camera_data()
  {
    auto message = plc_msgs::msg::CameraControlCustom();
    message.camera_id = this->get_parameter("camera_id").as_int();
    message.power = this->get_parameter("camera_power").as_bool();
    message.light = this->get_parameter("camera_light").as_bool();
    message.focus = this->get_parameter("camera_focus").as_double();
    message.zoom = this->get_parameter("camera_zoom").as_double();

    camera_publisher_->publish(message);
    RCLCPP_INFO(this->get_logger(), "Published Camera Data: [ID: %d, Power: %s, Light: %s, Focus: %.2f, Zoom: %.2f]",
                message.camera_id, message.power ? "true" : "false", message.light ? "true" : "false",
                message.focus, message.zoom);
  }

  void publish_lamp_data()
  {
    auto message = plc_msgs::msg::LampControlCustom();
    message.lamp_id = this->get_parameter("lamp_id").as_int();
    message.intensity = this->get_parameter("lamp_intensity").as_double();
    message.power = this->get_parameter("lamp_power").as_bool();

    lamp_publisher_->publish(message);
    RCLCPP_INFO(this->get_logger(), "Published Lamp Data: [ID: %d, Intensity: %.2f, Power: %s]",
                message.lamp_id, message.intensity, message.power ? "true" : "false");
  }

  void publish_panandtilt_data()
  {
    auto message = plc_msgs::msg::PanAndTiltControlCustom();
    message.panandtilt_id = this->get_parameter("panandtilt_id").as_int();
    message.x = this->get_parameter("panandtilt_x").as_double();
    message.y = this->get_parameter("panandtilt_y").as_double();
    message.z = this->get_parameter("panandtilt_z").as_double();

    panandtilt_publisher_->publish(message);
    RCLCPP_INFO(this->get_logger(), "Published Pan and Tilt Data: [ID: %d, X: %.2f, Y: %.2f, Z: %.2f]",
                message.panandtilt_id, message.x, message.y, message.z);
  }

  rclcpp::TimerBase::SharedPtr timer_camera_;
  rclcpp::TimerBase::SharedPtr timer_lamp_;
  rclcpp::TimerBase::SharedPtr timer_panandtilt_;
  rclcpp::Publisher<plc_msgs::msg::CameraControlCustom>::SharedPtr camera_publisher_;
  rclcpp::Publisher<plc_msgs::msg::LampControlCustom>::SharedPtr lamp_publisher_;
  rclcpp::Publisher<plc_msgs::msg::PanAndTiltControlCustom>::SharedPtr panandtilt_publisher_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ControlPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: ros2_publisher/plc_publisher/plc_publisher.py
import rclpy
from rclpy.node import Node
from plc_msgs.msg import CameraControlCustom, LampControlCustom, PanAndTiltControlCustom


class ControlPublisher(Node):
    def __init__(self):
        super().__init__('control_publisher')

        # Declare parameters for Camera Control
        self.declare_parameter('camera_id', 9)
        self.declare_parameter('camera_power', True)
        self.declare_parameter('camera_light', True)
        self.declare_parameter('camera_focus', 22.0)
        self.declare_parameter('camera_zoom', 25.0)

        # Declare parameters for Lamp Control
        self.declare_parameter('lamp_id', 9)
        self.declare_parameter('lamp_intensity', 80.0)
        self.declare_parameter('lamp_power', True)

        # Declare parameters for Pan and Tilt Control
        self.declare_parameter('panandtilt_id', 9)
        self.declare_parameter('panandtilt_x', 80.0)
        self.declare_parameter('panandtilt_y', 20.0)
        self.declare_parameter('panandtilt_z', 23.0)

        # Publishers for each topic
        self.camera_publisher = self.create_publisher(CameraControlCustom, 'CameraSetterTopic', 10)
        self.lamp_publisher = self.create_publisher(LampControlCustom, 'LampSetterTopic', 10)
        self.panandtilt_publisher = self.create_publisher(PanAndTiltControlCustom, 'PanAndTiltSetterTopic', 10)

        # Timers for periodic publishing (1 Hz frequency)
        self.create_timer(1.0, self.publish_camera_data)
        self.create_timer(1.0, self.publish_lamp_data)
        self.create_timer(1.0, self.publish_panandtilt_data)

    def publish_camera_data(self):
        msg = CameraControlCustom()
        msg.cameraID = int(self.get_parameter('camera_id').value)
        msg.power = bool(self.get_parameter('camera_power').value)
        msg.light = bool(self.get_parameter('camera_light').value)
        msg.focus = float(self.get_parameter('camera_focus').value)
        msg.zoom = float(self.get_parameter('camera_zoom').value)

        self.camera_publisher.publish(msg)
        self.get_logger().info(f'Published Camera Data: {msg}')

    def publish_lamp_data(self):
        msg = LampControlCustom()
        msg.lampID = int(self.get_parameter('lamp_id').value)
        msg.intensity = float(self.get_parameter('lamp_intensity').value)
        msg.power = bool(self.get_parameter('lamp_power').value)

        self.lamp_publisher.publish(msg)
        self.get_logger().info(f'Published Lamp Data: {msg}')

    def publish_panandtilt_data(self):
        msg = PanAndTiltControlCustom()
        msg.panandtiltID = int(self.get_parameter('panandtilt_id').value)
        msg.x = float(self.get_parameter('panandtilt_x').value)
        msg.y = float(self.get_parameter('panandtilt_y').value)
        msg.z = float(self.get_parameter('panandtilt_z').value)

        self.panandtilt_publisher.publish(msg)
        self.get_logger().info(f'Published Pan and Tilt Data: {msg}')


def main(args=None):
    rclpy.init(args=args)

    control_publisher = ControlPublisher()

    try:
        rclpy.spin(control_publisher)
    except KeyboardInterrupt:
        control_publisher.get_logger().info("Shutting down...")
    finally:
        control_publisher.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
